devito/profiler.py

```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
                            "(double)(end_%(n)s.tv_sec-start_%(n)s.tv_sec)+" +
                            "(double)(end_%(n)s.tv_usec-start_%(n)s.tv_usec)" +
                            "/1000000") %
-                          {"sn": self.name, "n": name}),
+                          {"sn": self.t_name, "n": name}),
             ])
         ]
 
-        return init + block + end
+        return init + code + end
 
-    @property
-    def get_timings_class(self):
+    def get_oi_and_flops(self, name, code, size):
+        """Calculates the total operation intensity of the code provided.
+        If needed, lets the C code calculate it.
+
+        :param name: The name of the field to be populated
+        :param code: The code to be profiled.
+        """
+        if name not in self.oi:
+            self.oi[name] = 0
```

Comment:
If you use a ``defaultdict``, an entry which is not in the dict gets assigned a default value (0 in the case of int). Look at this:

```
In [1]: from collections import defaultdict

In [2]: d = defaultdict(int)

In [3]: d[1]
Out[3]: 0

In [4]: d[2] += 1

In [5]: d
Out[5]: defaultdict(<type 'int'>, {1: 0, 2: 1})
```

Similar behaviour with a normal dict and using the ``setdefault`` method ({}.setdefault.__doc__ is quite self-explanatory)